File: cpp/src/hypocenter.cpp
#include "hypocenter.h"

#include <limits>

// JSON Keys
#define LATITUDE_KEY "Latitude"
#define LONGITUDE_KEY "Longitude"
#define DEPTH_KEY "Depth"
#define TIME_KEY "Time"
#define LATITUDE_ERROR_KEY "LatitudeError"
#define LONGITUDE_ERROR_KEY "LongitudeError"
#define DEPTH_ERROR_KEY "DepthError"
#define TIME_ERROR_KEY "TimeError"

namespace detectionformats {

hypocenter::hypocenter() {
	latitude = std::numeric_limits<double>::quiet_NaN();
	longitude = std::numeric_limits<double>::quiet_NaN();
	depth = std::numeric_limits<double>::quiet_NaN();
	time = std::numeric_limits<double>::quiet_NaN();
	latitudeerror = std::numeric_limits<double>::quiet_NaN();
	longitudeerror = std::numeric_limits<double>::quiet_NaN();
	deptherror = std::numeric_limits<double>::quiet_NaN();
	timeerror = std::numeric_limits<double>::quiet_NaN();

}

hypocenter::hypocenter(double newlatitude, double newlongitude, double newtime,
		double newdepth, double newlatitudeerror, double newlongitudeerror,
		double newtimeerror, double newdeptherror) {
	latitude = newlatitude;
	longitude = newlongitude;
	depth = newdepth;
	time = newtime;
	latitudeerror = newlatitudeerror;
	longitudeerror = newlongitudeerror;
	deptherror = newdeptherror;
	timeerror = newtimeerror;
}

hypocenter::hypocenter(rapidjson::Value &json) {
	// required values
	// latitude
	if ((json.HasMember(LATITUDE_KEY) == true)
			&& (json[LATITUDE_KEY].IsNumber() == true)
			&& (json[LATITUDE_KEY].IsDouble() == true))
		latitude = json[LATITUDE_KEY].GetDouble();
	else
		latitude = std::numeric_limits<double>::quiet_NaN();

	// longitude
	if ((json.HasMember(LONGITUDE_KEY) == true)
			&& (json[LONGITUDE_KEY].IsNumber() == true)
			&& (json[LONGITUDE_KEY].IsDouble() == true))
		longitude = json[LONGITUDE_KEY].GetDouble();
	else
		longitude = std::numeric_limits<double>::quiet_NaN();

	// time
	if ((json.HasMember(TIME_KEY) == true)
			&& (json[TIME_KEY].IsString() == true))
		time = detectionformats::ConvertISO8601ToEpochTime(
				std::string(json[TIME_KEY].GetString(),
						json[TIME_KEY].GetStringLength()));
	else
		time = std::numeric_limits<double>::quiet_NaN();

	// depth
	if ((json.HasMember(DEPTH_KEY) == true)
			&& (json[DEPTH_KEY].IsNumber() == true)
			&& (json[DEPTH_KEY].IsDouble() == true))
		depth = json[DEPTH_KEY].GetDouble();
	else
		depth = std::numeric_limits<double>::quiet_NaN();

	// optional values
	// latitude error
	if ((json.HasMember(LATITUDE_ERROR_KEY) == true)
			&& (json[LATITUDE_ERROR_KEY].IsNumber() == true)
			&& (json[LATITUDE_ERROR_KEY].IsDouble() == true))
		latitudeerror = json[LATITUDE_ERROR_KEY].GetDouble();
	else
		latitudeerror = std::numeric_limits<double>::quiet_NaN();

	// longitude error
	if ((json.HasMember(LONGITUDE_ERROR_KEY) == true)
			&& (json[LONGITUDE_ERROR_KEY].IsNumber() == true)
			&& (json[LONGITUDE_ERROR_KEY].IsDouble() == true))
		longitudeerror = json[LONGITUDE_ERROR_KEY].GetDouble();
	else
		longitudeerror = std::numeric_limits<double>::quiet_NaN();

	// time error
	if ((json.HasMember(TIME_ERROR_KEY) == true)
			&& (json[TIME_ERROR_KEY].IsNumber() == true)
			&& (json[TIME_ERROR_KEY].IsDouble() == true))
		timeerror = json[TIME_ERROR_KEY].GetDouble();
	else
		timeerror = std::numeric_limits<double>::quiet_NaN();

	// depth error
	if ((json.HasMember(DEPTH_ERROR_KEY) == true)
			&& (json[DEPTH_ERROR_KEY].IsNumber() == true)
			&& (json[DEPTH_ERROR_KEY].IsDouble() == true))
		deptherror = json[DEPTH_ERROR_KEY].GetDouble();
	else
		deptherror = std::numeric_limits<double>::quiet_NaN();
}

hypocenter::hypocenter(const hypocenter & newhypocenter) {
	latitude = newhypocenter.latitude;
	longitude = newhypocenter.longitude;
	depth = newhypocenter.depth;
	time = newhypocenter.time;
	latitudeerror = newhypocenter.latitudeerror;
	longitudeerror = newhypocenter.longitudeerror;
	deptherror = newhypocenter.deptherror;
	timeerror = newhypocenter.timeerror;
}

hypocenter::~hypocenter() {
}

rapidjson::Value & hypocenter::tojson(rapidjson::Value &json,
		rapidjson::MemoryPoolAllocator<rapidjson::CrtAllocator> &allocator) {
	json.SetObject();

	// required values
	// latitude
	if (std::isnan(latitude) != true)
		json.AddMember(LATITUDE_KEY, latitude, allocator);

	// longitude
	if (std::isnan(longitude) != true)
		json.AddMember(LONGITUDE_KEY, longitude, allocator);

	// time
	if (std::isnan(time) != true) {
		std::string timestring = detectionformats::ConvertEpochTimeToISO8601(
				time);
		rapidjson::Value timevalue;
		timevalue.SetString(rapidjson::StringRef(timestring.c_str()),
				allocator);
		json.AddMember(TIME_KEY, timevalue, allocator);
	}

	// depth
	if (std::isnan(depth) != true)
		json.AddMember(DEPTH_KEY, depth, allocator);

	// optional values
	// latitude error
	if (std::isnan(latitudeerror) != true)
		json.AddMember(LATITUDE_ERROR_KEY, latitudeerror, allocator);

	// longitude error
	if (std::isnan(longitudeerror) != true)
		json.AddMember(LONGITUDE_ERROR_KEY, longitudeerror, allocator);

	// time error
	if (std::isnan(timeerror) != true)
		json.AddMember(TIME_ERROR_KEY, timeerror, allocator);

	// depth error
	if (std::isnan(deptherror) != true)
		json.AddMember(DEPTH_ERROR_KEY, deptherror, allocator);

	return (json);
}

std::vector<std::string> hypocenter::geterrors() {
	std::vector<std::string> errorlist;

	// check required data
	// latitude
	if (std::isnan(latitude) == true) {
		// latitude not found
		errorlist.push_back("No Latitude in hypocenter class.");
	} else if ((latitude < -90) || (latitude > 90)) {
		errorlist.push_back("Invalid Latitude in hypocenter class.");
	}

	// longitude
	if (std::isnan(longitude) == true) {
		// longitude not found
		errorlist.push_back("No Longitude in hypocenter class.");
	} else if ((longitude < -180) || (longitude > 180)) {
		errorlist.push_back("Invalid Longitude in hypocenter class.");
	}

	// time
	if (std::isnan(time) == true) {
		errorlist.push_back("Time is missing in hypocenter class.");
	} else {
		try {
			if (detectionformats::IsStringISO8601(
					detectionformats::ConvertEpochTimeToISO8601(time))
					== false) {
				errorlist.push_back(
						"Time did not validate in hypocenter class.");
			}
		} catch (const std::exception & e) {
			errorlist.push_back(std::string(e.what()));
		}
	}

	// depth
	if (std::isnan(depth) == true) {
		// depth not found
		errorlist.push_back("No Depth in hypocenter class.");
	} else if ((depth < -100) || (depth > 1500)) {
		errorlist.push_back("Invalid Depth in hypocenter class.");
	}

	// optional keys
	// Currently no validation criteria for optional values LatitudeError,
	// LongitudeError, TimeError, and DepthError.

	// return the list of errors
	return (errorlist);
}

}

File: python/detectionformats/hypocenter.py
#!/usr/bin/env python

# stdlib imports
import json
import datetime


class Hypocenter:
    """Hypocenter - a conversion class used to create, parse, and validate
    hypocenter data as part of detection data.
    """

    # json keys
    LATITUDE_KEY = "Latitude"
    LONGITUDE_KEY = "Longitude"
    DEPTH_KEY = "Depth"
    TIME_KEY = "Time"
    LATITUDE_ERROR_KEY = "LatitudeError"
    LONGITUDE_ERROR_KEY = "LongitudeError"
    DEPTH_ERROR_KEY = "DepthError"
    TIME_ERROR_KEY = "TimeError"

    def __init__(
        self,
        newLatitude=None,
        newLongitude=None,
        newDepth=None,
        newTime=None,
        newLatitudeError=None,
        newLongitudeError=None,
        newDepthError=None,
        newTimeError=None,
    ):
        """Initialize the hypocenter object. Constructs an empty object
           if all arguments are None

        Args:
            newLatitude: a required Number containing the latitude as a float in
                degrees
            newLongitude: a required Number containing the longitude as a float
                in degrees
            newDepth: a required Number containing the depth as a float in km
            newTime: a required datetime containing the correlation time
            newMagnitude: an optional Number containing the desired magnitude
                estimate as a float
            newLatitudeError: an optional Number containing the latitude error
                measurement as a float
            newLongitudeError: an optional Number containing the longitude error
                measurement as a float
            newDepthError: an optional Number containing the depth error
                measurement as a float
            newTimeError: an optional Number containing the time error measurement
                as a float
        Returns:
            Nothing
        Raises:
            Nothing
        """
        # first required keys
        if newLatitude is not None:
            self.latitude = newLatitude
        if newLongitude is not None:
            self.longitude = newLongitude
        if newDepth is not None:
            self.depth = newDepth
        if newTime is not None:
            self.time = newTime

        # second optional keys
        if newLatitudeError is not None:
            self.latitudeError = newLatitudeError

        if newLongitudeError is not None:
            self.longitudeError = newLongitudeError

        if newDepthError is not None:
            self.depthError = newDepthError

        if newTimeError is not None:
            self.timeError = newTimeError

    def fromJSONString(self, jsonString):
        """Populates the object from a json formatted string

        Args:
            jsonString: a required String containing the json formatted text
        Returns:
            Nothing
        Raises:
            Nothing
        """
        jsonObject = json.loads(jsonString)
        self.fromDict(jsonObject)

    def fromDict(self, aDict):
        """Populates the object from a dictionary

        Args:
            aDict: a required dictionary
        Returns:
            Nothing
        Raises:
            Nothing
        """
        # first required keys
        try:
            self.latitude = aDict[self.LATITUDE_KEY]
            self.longitude = aDict[self.LONGITUDE_KEY]
            self.depth = aDict[self.DEPTH_KEY]
            timeString = aDict[self.TIME_KEY][:-1] + "000Z"
            self.time = datetime.datetime.strptime(timeString, "%Y-%m-%dT%H:%M:%S.%fZ")
        except (ValueError, KeyError, TypeError) as e:
            print("Dict format error, missing required keys: %s" % e)

        # second optional keys
        if self.LATITUDE_ERROR_KEY in aDict:
            self.latitudeError = aDict[self.LATITUDE_ERROR_KEY]

        if self.LONGITUDE_ERROR_KEY in aDict:
            self.longitudeError = aDict[self.LONGITUDE_ERROR_KEY]

        if self.DEPTH_ERROR_KEY in aDict:
            self.depthError = aDict[self.DEPTH_ERROR_KEY]

        if self.TIME_ERROR_KEY in aDict:
            self.timeError = aDict[self.TIME_ERROR_KEY]

    def toJSONString(self):
        """Converts the object to a json formatted string

        Args:
            None
        Returns:
            The JSON formatted message as a String
        Raises:
            Nothing
        """
        jsonObject = self.toDict()

        return json.dumps(jsonObject, ensure_ascii=False)

    def toDict(self):
        """Converts the object to a dictionary

        Args:
            None
        Returns:
            The dictionary
        Raises:
            Nothing
        """
        aDict = {}

        # first required keys
        try:
            aDict[self.LATITUDE_KEY] = self.latitude
            aDict[self.LONGITUDE_KEY] = self.longitude
            aDict[self.DEPTH_KEY] = self.depth
            timeString = self.time.isoformat(timespec="milliseconds") + "Z"
            aDict[self.TIME_KEY] = timeString
        except (NameError, AttributeError) as e:
            print("Missing required data error: %s" % e)

        # second optional keys
        if hasattr(self, "latitudeError"):
            aDict[self.LATITUDE_ERROR_KEY] = self.latitudeError

        if hasattr(self, "longitudeError"):
            aDict[self.LONGITUDE_ERROR_KEY] = self.longitudeError

        if hasattr(self, "depthError"):
            aDict[self.DEPTH_ERROR_KEY] = self.depthError

        if hasattr(self, "timeError"):
            aDict[self.TIME_ERROR_KEY] = self.timeError

        return aDict

    def isValid(self):
        """Checks to see if the object is valid

        Args:
            None
        Returns:
            True if the object is valid, False otherwise
        Raises:
            Nothing
        """
        errorList = self.getErrors()

        return not errorList

    def getErrors(self):
        """Gets a list of object validation errors

        Args:
            None
        Returns:
            A List of Strings containing the validation error messages
        Raises:
            Nothing
        """
        errorList = []

        # required values
        try:
            if self.latitude < -90 or self.latitude > 90:
                errorList.append(
                    "Latitude in Hypocenter Class not in the range of -90 to 90."
                )
        except (NameError, AttributeError):
            errorList.append("No Latitude in Hypocenter Class.")

        try:
            if self.longitude < -180 or self.longitude > 180:
                errorList.append(
                    "Longitude in Hypocenter Class not in the range of -180 to 180."
                )
        except (NameError, AttributeError):
            errorList.append("No Longitude in Hypocenter Class.")

        try:
            if self.depth < -100 or self.depth > 1500:
                errorList.append(
                    "Depth in Hypocenter Class not in the range of -100 to 1500."
                )
        except (NameError, AttributeError):
            errorList.append("No Depth in Hypocenter Class.")

        return errorList
